Remove Serving Node from architecture (#27)
mlops-labs-team1/engineering.labs: Lab1_Operationalizing_Pytorch_with_Mlflow/README.md

| 8 | Project Metadata | We need to describe the ML Component, its structure and how it should run | Source Code Repository<br>Training Node | MLFlow Project Definition | Non-functional Requirement we must adhere |
| 9 | ML Lifecycle Management | We must track the ML Process | All Nodes and Machines | MLFlow |  Non-functional Requirement we must adhere |
| 10 | ML Framework | We are building ML Components. Therefore, we must use a ML framework to train our models | Training Node<br>Serving Instances | Pytorch | Non-functional Requirement we must adhere |

### Workflow

[DevOps](https://en.wikipedia.org/wiki/DevOps) tools heavily relies on actions triggered by changes
and other events regarding application's source code. MLOps goes beyond this assumption since any reasonable
solution must deal with assets like ML Models, datasets, features, etc. Therefore, a simples ML 
application may turn into a complex workflow in order to ensure the best model runs in Production. The
following workflow depicts a common development scenario where a production web application uses a ML
Model to provide business value. 

![Solution Workflow](imgs/workflow.svg)

> The first stage is Development Workflow. It concerns the continuous integration of new ML code and Dataset 
> changes into a Development Server. Here follows the steps:

Reviewer: Does it makes sense here or before here to talk about the experimentation work that can be done in MLflow before we get to the point of productionisation? I only mention it since the strong part of MLflow is that a (potentially cowboy!) data scientist can do all their experimentation work in MLflow, track parameters, compare the performance etc. Then they decide what to commit to master, and the MLops pipeline we've built out here takes over, but lets them see the progress in their "own" tool.
```suggestion
Before this MLops workflow is used, there would be an experimentation workflow used by a data scientist. They would be able to use this repository to test out various models, architectures, hyperparameters etc. which would all be logged and tracked in MLflow available for them to compare the performance in the UI. Once they had decided on a new model to move to production, they can create a Pull Request, ready for their team to review, which would then move into the MLops workflow proper.
The first stage is Development Workflow. It concerns the continuous integration of new ML code and Dataset 
changes into a Development Server. Here follows the steps:
```

Author: John... you're totally right. It's a great suggestion. I thought it would be nice if we could also mention this during the context and give MLFlow a nice spot. So, I did quite a big change in the text. Don't know if its too formal for our purposes. Anyway, take a look and if you think it doesn't fit there we may rollback it and accept your text right away. :wink:

1. Commited code or changes in the Dataset fires a training execution. These events trigger Github 
Actions (GH Actions);
 
 1. ML pipeline starts. we are considering that previous data activities (Extraction, Validation and
 Preparation) were already executed. GH Actions issues a _gcloud_ command to _Training Node_, so the 
 ML Component code gets up to date. The ML Component itself describes its structure (_MLFlow project file_) 
 and how it should be run (_Dockerfile_). The MLFlow CLI starts the training process by issuing: 
 ```mlflow run```. The training happens in a GPU-powered Docker container configured with the 
 _Tracking Server_ (```MLFLOW_TRACKING_URI```) . At the end of this step, there's a new trained model 
 with its respective metadata;
 
 1. The Training Node sends the trained model, its extra files, and metadata to the Model Registry
 and Tracking Server respectively. In fact, these assets are physically stored in a Cloud
 Storage Service acessible from Internet;
 
 1. The development workflow ends by deploying the trained model in Google Cloud Run Service. 

>  GH Actions creates a new image which contains the trained model (```mlflow create deployment```) which

Reviewer: ```suggestion
 GH Actions creates a new image which contains the trained model (```mlflow create deployment```) which is tagged with the MLflow model name and version and then
```

>  is pushed to Google Container Registry.

Reviewer: ```suggestion
 is pushed to Google Container Registry. This image is deployed to a development Cloud Run deployment, where various acceptance tests can be performed to ensure the image is ready for production.
```

The decision to promote a model to Production happens after some experiments. Model version _X_ is selected. This _external event_ triggers the Production Workflow:

> 5. Operations and Product Management Team decides to deploy a model to Production environment. A release command is issued, triggering GitHub Actions. The runner pulls and configures the corresponding Docker image according to Production environment;

Reviewer: ```suggestion
5. Operations and Product Management Team decides to deploy a model to Production environment. A release command is issued, specifying a model name and version to deploy, triggering GitHub Actions. The runner pulls and configures the corresponding Docker image and deploys this to the Production Cloud Run environment
```

# Operationalizing Pytorch Model with Mlflow

## Context

PyTorch is one of the main machine learning libraries used in applications such as computer vision and 
natural language processing. It is catching up with TensorFlow accelerating the path from research 
prototyping to production deployment. Recently, Pytorch community announced a number of technical 
contributions to enable end-to-end support for MLflow usage with PyTorch. 

MLflow is an open source platform to manage the ML lifecycle, including experimentation, reproducibility, deployment, and a central model registry. One may interact with its components by command line interface
or well-known APIs (Python, R, Java and REST), building a workflow capable of handling development and
production activies.

In this lab, [MLOps community](https://mlops.community/) wants to test them and provide feasible ways 
to build end-to-end model lifecycles of Pytorch assets using Mlflow.

## Team

- Alexey Naiden
- John Savage
- Michel Vasconcelos
- Varuna Jayasiri

## Repo Organization

Following list describes the most relevant folders in which this lab is organized. Some folders in 

> this repo are self-explanatory (like [_imgs_](imgs/)) and therefore aren'r mentioned in the list.

Reviewer 2: super minor: typo, `aren'r`

Author: Done!

The following figure depicts the proposed architecture.

![Proposed architecture](imgs/arch.svg)

**Nodes** are bare-metal or virtual machines that host a part of the pipeline process. **ML components** 
are ML-based code that should be trained, tested and deployed by end-to-end pipeline.They must be 
packaged as [MLFlow projects](https://mlflow.org/docs/latest/projects.html). **Training nodes**
host the training process which runs in **Training containers**, running dockerized images built 
with MLFlow and necessary libs to train ML components. Each ML component defines its needs, that is, 
libs and external dependencies. To ensure traceability, provenance and model configuration management, 
one may track hyperparameters, training steps, test results, extra data and the model itself into 
_MLFlow Tracking Server_ hosted in the **Tracking Node**. All these items, metadata and trained models, 
are stored in external resources: RDBMS and Cloud Storage Services. Trained models are embedded into 
dockerized **Serving Instances**, containers running [Torchserve](https://pytorch.org/serve/) providing 

> web access to the models.

Reviewer 2: minor: I would say `providing access to the models via HTTP`, because there's no web UI per se.

Author: Yes... pretty much better! ;-)

In order to provide feedback loop, a **Monitor Node** checks the status and performance meters, issuing 

> new training cycles or a full rebuild process. Some structures traverses through Development and Production perimeters. Training and Monitor nodes are key elements that communicate between those zones. 

Reviewer 2: > Some structures traverses through Development and Production perimeters

I don't get this phrase. Would you care to elaborate?

Author: Usually, when we think about "system architecture" we try to isolate elements in order to comply with security requirements. Something like Development traffic never reaches Production zone, that is, Development nodes would "lie" in Development zone with no connectivity with nodes \ servers in Production. However, when I drew that, I was thinking a "peculiar" Monitoring Node. One that could reach Development and Production elements. It would lie in between zones. Just like a "cross-cutting concern". 

I was trying to (concisely) say that with that phrase. However, my summarization skills are flawed somehow. hahahahahaahahaha 

In fact, maybe we shouldn't explain that at all. Just remove that keep it simple. What do you think? :-)

Author: I have removed it. ;-)

|---|---------|-------------|----------------|----------|-----------|
| 1 | Computing | ML needs processing power as well as GPU. Our architecure relies on distributed elements, thus needing a reasonable number of physical or virtual machines (some with GPU-power) | All Nodes and Machines | GCP Compute Service | It provides $ 300,00 of (enough) free services and we can also spin some gpu-powered machines |
| 2 | Durable Data Storage | Every serious solution needs to store durable data | Tracking Server | Google Cloud SQL | A Postgres Database hosted by Google Cloud SQL is reliable and fast enough for our needs  |
| 3 | Large and Binary Data Storage | ML projects eventually need to store models and extra large files, This type of data doesn't fit in RDBMS | Tracking Server<br>Artifact Store | Google Cloud Storage | Same as #1 |
| 4 | Shipment and Deployment | There are many deployable assets as well as acessory tools that need specific environment and SO libraries to run. Just installing them into a machine isn't a viable option | Training  Node<br>Google Cloud Services | Docker | It's the _de facto_ pattern for shipping and deploying things |

> | 5 | Lib Management | Python has a powerful but sometimes confusing and conflicting library ecossystem. Our components may depend on conflicting libs and that will lead to problems in Production | All Nodes and Machines | Conda<br>Pip | Both are heavily used by Python community. They also work quite well with isolation provided by Docker images |

Reviewer 2: We don't use Conda anymore, do we?

Author: Oooops. Removed!

| 6 | Remote Access | Sometimes we need to issue remote commands or operate a remote machine | All Nodes and Machines | SSH | Hey... It's SSH. There's no need to rationale :-P |
| 7 | CI/CD & Workflow Management | We need to seamlessly integrate and deploy ML stuff. We also have to coordinate the process between nodes | All Items | Github Actions | We're using GitHub, it's heavily used by community and comes off-the-shelf |

> | 8 | Scheduler | Some activities are triggered by clock | All Nodes and Machines | CRON | We are using Unix like machines |

Reviewer 2: Just wondering what we use cron for

Author: Nothing. Forgot it there. We thought it would be of use, but we can do everything with GH + gcloud.

| 11 | ML Framework | We are building ML Components. Therefore, we must use a ML framework to train our models | Training Node<br>Serving Instances | Pytorch | Non-functional Requirement we must adhere |

### Workflow

[DevOps](https://en.wikipedia.org/wiki/DevOps) tools heavily relies on actions triggered by changes
and other events regarding application's source code. MLOps goes beyond this assumption since any reasonable
solution must deal with assets like ML Models, datasets, features, etc. Therefore, a simples ML 
application may turn into a complex workflow in order to ensure the best model runs in Production. The
following workflow depicts a common development scenario where a production web application uses a ML
Model to provide business value. 

![Solution Workflow](imgs/workflow.svg)

The first stage is Development Workflow. It concerns the continuous integration of new ML code and Dataset 
changes into a Development Server. Here follows the steps:

1. Commited code or changes in the Dataset fires a training execution. These events trigger Github 
Actions (GH Actions);
 
 1. ML pipeline starts. we are considering that previous data activities (Extraction, Validation and

>  Preparation) were already executed. GH Actions issues a _gcloud commandIn to _Training Node_, so the 

Reviewer 2: minor: I think it should be `_gcloud_ command to _Training Node_`. Or something like that?

Author: Forgot that. Fixed!

solution must deal with assets like ML Models, datasets, features, etc. Therefore, a simple ML 
application may turn into a complex workflow in order to ensure the best model runs in Production.
Besides that, developing ML models demand a relevant time assessing and comparing parameters, algorithms
and model performance. MLOps goal is to address these concerns by tying tools and ML frameworks in order to quickly 
deploy models to Production (and keep them there).

PyTorch is one of the main machine learning libraries used in applications such as computer vision and 
natural language processing. It is catching up with TensorFlow accelerating the path from research 
prototyping to production deployment. Recently, Pytorch community announced a number of technical 
contributions to enable end-to-end support for MLflow usage with PyTorch. 

MLflow is an open source platform to manage the ML lifecycle, including experimentation, reproducibility, 
deployment, and a central model registry. One may interact with its components by command line interface
or well-known APIs (Python, R, Java and REST), building a workflow capable of handling development and

> production activies. Actually, Its simple design allows small teams (or a lone cowboy data scientist as well) 

Reviewer: Ha, cowboy might be a bit insulting out of context!!
```suggestion
production activies. Actually, Its simple design allows small teams (or a lone data scientist as well) 
```

Author: "Cowboy" removed. Still wanna be a cowboy something... maybe bc old western movies... haha
